### PR TITLE
    datamigration: package created

### LINKED ISSUE #1048: Implement data migration
As IOV I want to be able to write a data migration that will be executed exactly once in the future.

We are running `bnsd` on production, and we can no longer rely on the genesis to initialize state. 
Migration is not a problem when we migrate an existing data, because we can use [`migration` package](https://github.com/iov-one/weave/blob/master/migration/doc.go). 
Migration is currently impossible when a new functionality is introduced, or we wish to create or delete data. This is not possible with the existing `migration` package.

A proposed solution is to implement a new module where we define state modification. Execution of defined state modifications is triggered by a message.

Here is my earlier proof of concept implementation: https://github.com/iov-one/weave/commit/42f966e2213fc0dc244edce31152980728cd53fa
Data migration is defined as a Go function in the application that can access database 

https://github.com/iov-one/weave/blob/42f966e2213fc0dc244edce31152980728cd53fa/cmd/bnsd/x/statefix/statefixes.go#L7-L14

## Acceptance criteria
- [x] Data migration is a Go function, written as part of the application that execution cannot be altered or influenced from the outside,
- [x] Data migration can access database and change its state.
- [x] Each data migration can be executed only once. Execution of a state migration is stored on the blockchain, so that replying is possible,
- [ ] Anyone can request execution of a state migration,

### PR DESCRIPTION
**Pull request by husio**
A new package `datamigration` was created.  `datamigration` allows to
    create and execute data modification on chain. Each migration is
    declared as a Go code, that can access and modify database state.
    
    This functionality is helpful when migrating state of an existing code
    to a new implementation or when an initialization of a newly installed
    extension is needed, but the genesis cannot be used.

I have registered a data migration for the `x/msgfee` package to initialize the configuration owner address if not set. https://github.com/iov-one/weave/pull/1056/commits/a28f13480f178596029f84c50d808c4998ba1fba#diff-8e2353da9163d2767044318c783c22f0R19-R43

Testing locally:
```
$ bnscli upgrade-schema -pkg datamigration -ver 1 \
    | bnscli with-fee -amount 1IOV \
    | bnscli sign \
    | bnscli submit 

$ bnscli datamigration -id "initialize x/msgfee configuration owner" \
    | bnscli with-fee -amount 1IOV \
    | bnscli sign \
    | bnscli submit 
```
    
resolve #1048

### CODE REVIEW

File: datamigration/codec.proto [hunk 1]
```diff
@@ -0,0 +1,15 @@
+syntax = "proto3";
+
+package datamigration;
+
+import "codec.proto";
+import "gogoproto/gogo.proto";
+
+message ExecutedMigration {
+  weave.Metadata metadata = 1;
```
Comment (orkunkl) [inline datamigration/codec.proto]: Can you explain to me why there is no data here? How this model represents an executed migration?

File: datamigration/doc.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
+/*
+
+Data migration package implements a functionality to declare and execute data
+migrations.
+
+Each data migration is a function registered under a unique name, that can be
+executed only once.
+
+In order to register a new data migration:
```
Comment (orkunkl) [inline datamigration/doc.go]: I found this _steps_ to be insufficient. For example **register a new migration using datamigration.MustRegister** but register in where? in the `cmd/app` folder?

File: datamigration/doc.go [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+1. register a new migration using `datamigration.MustRegister`,
+2. build a new binary of your application and make sure all validators are running it,
+3. send a transaction with ExecuteMigrationMsg, signed by required addresses,
+4. if migration was executed, it became a part of the chain. Its code must
```
Comment (orkunkl) [inline datamigration/doc.go]: What it means here **it became a part of the chain**? Do you mean `migration result becomes persistent in the chain and irreversible`? And what do you mean by **its code must remain unchanged forever**?

File: datamigration/handler_test.go [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+					RequiredSigners: []weave.Address{aliceCond.Address()},
+					Migrate:         func(context.Context, weave.KVStore) error { return nil },
+				},
+			)
+			MustRegister(
+				"second migration",
+				Migration{
+					ChainID:         "staging",
+					RequiredSigners: []weave.Address{aliceCond.Address()},
+					Migrate:         func(context.Context, weave.KVStore) error { return nil },
```
Comment (orkunkl) [inline datamigration/handler_test.go]: Can you write a mock migration function test that actually migrates data to demonstrate how to write functions?

File: datamigration/codec.proto (hunk 1 above)
Comment (husio) [reply datamigration/codec.proto]: There is no data that I need to store. I only need to keep an information that a migration was executed. Value of the stored pair does not matter.

File: datamigration/doc.go (hunk 2 above)
Comment (husio) [reply datamigration/doc.go]: I will update this description. The current instruction expects that you have set everything up before and now you only want to add another migration.

File: datamigration/doc.go (hunk 3 above)
Comment (husio) [reply datamigration/doc.go]: > it became a part of the chain

Blockchain is a data structure. In a way, it is a memory of a database. Maybe this is too big of a mind jump.
The important part is that migration is persisted on the chain. Maybe that is what I should write there.

> its code must remain unchanged forever

Exactly what I wrote :) You must never change or delete the code of a migration.

File: datamigration/handler_test.go (hunk 4 above)
Comment (husio) [reply datamigration/handler_test.go]: I think this is outside of the scope of this functionality. I give you the database and it is up to the user to decide what to do. You will probably use a bucket to create/update/delete entities.

File: cmd/bnsd/app/datamigration.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+	"github.com/iov-one/weave"
+	"github.com/iov-one/weave/datamigration"
+	"github.com/iov-one/weave/errors"
+	"github.com/iov-one/weave/gconf"
+	"github.com/iov-one/weave/x/msgfee"
+)
+
+func init() {
+	admin, err := weave.ParseAddress("cond:gov/rule/0000000000000003")
```
Comment (husio) [inline cmd/bnsd/app/datamigration.go]: @davepuchyr can you confirm that this should be the admin of the `msgfee` extension? Admin can modify configuration further and set for example who is allowed to change prices.

Comment (davepuchyr) [reply cmd/bnsd/app/datamigration.go]: Confirmed.

**Pull request merged**